addons/stock/models/stock_inventory.py

```diff
@@ -179,6 +192,9 @@ def _get_inventory_lines_values(self):
         locations = self.env['stock.location'].search([('id', 'child_of', [self.location_id.id])])
         domain = ' location_id in %s'
         args = (tuple(locations.ids),)
+        vals = []
+        Product = products = self.env['product.product']
+        quant_prods = Product    
```

Comment:
You are right in your last test with example, you were assigning to a[0], So, from the point of view of a[0], you're rebinding a[0], not value of a. 
Here In Our case we are reassigning Products and QuantProducts with different value which does not effect each other.
For Example :
k = l = []
k = [1,2,3,4]
l = [5,6,7]
here we assiging k to differente value which doen not change l value. same in our case products and quantproducts does not effect eachother.
second example : 
 
d=e=f=3
e=4
print('f:',f)
print('e:',e)
Here, we're rebinding the name e to the value 4. That doesn't affect the names d and f in any way.